File: src/models/testCase.js
// /src/models/testCase.js
/**
 * 테스트케이스 데이터 모델
 * 폴더(테스트 그룹)와 테스트케이스를 동일한 구조로 관리
 */

// 기본 테스트케이스 객체 생성 함수
export const createTestCase = (id, name, parentId = null, type = 'testcase', description = '') => ({
    id,
    name,
    parentId,
    type, // 'folder' 또는 'testcase'
    description,
    steps: [],
    expectedResults: '',
    createdAt: new Date().toISOString(),
    updatedAt: new Date().toISOString()
  });
  
  // 테스트케이스 폴더(그룹) 생성 함수
  export const createTestFolder = (id, name, parentId = null) => ({
    id,
    name,
    parentId,
    type: 'folder',
    createdAt: new Date().toISOString(),
    updatedAt: new Date().toISOString()
  });
  
  // 테스트 단계 객체 생성 함수
  export const createTestStep = (stepNumber, description = '', expectedResult = '') => ({
    stepNumber,
    description,
    expectedResult
  });
  
  // 초기 테스트케이스 샘플 데이터
  export const initialTestCases = [
    // 폴더 5개 생성
    createTestFolder("folder-1", "Test Folder 1"),
    createTestFolder("folder-2", "Test Folder 2"),
    createTestFolder("folder-3", "Test Folder 3"),
    createTestFolder("folder-4", "Test Folder 4"),
    createTestFolder("folder-5", "Test Folder 5"),
    // 각 폴더에 20개씩 테스트 케이스 생성
    ...Array.from({ length: 100 }, (_, i) => {
      const folderIdx = Math.floor(i / 20) + 1;
      return createTestCase(
        `test-${i + 1}`,
        `테스트 케이스 ${i + 1}`,
        `folder-${folderIdx}`,
        "testcase",
        `테스트 케이스 ${i + 1}의 설명입니다.`,
        [
          createTestStep(1, "Step 1 설명", "Step 1 기대 결과"),
          createTestStep(2, "Step 2 설명", "Step 2 기대 결과"),
        ],
        "전체 기대 결과"
      );
    }),
  ];

File: backend_for_testcase/src/main/java/com/testcase/testcasemanagement/model/TestCase.java
// src/main/java/com/testcase/testcasemanagement/model/TestCase.java

package com.testcase.testcasemanagement.model;

import com.fasterxml.jackson.annotation.JsonBackReference;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.GenericGenerator;

import java.time.LocalDateTime;
import java.util.List;

@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(
        name = "testcases",
        uniqueConstraints = {
                @UniqueConstraint(columnNames = {"project_id", "name", "parent_id", "type"}),
                @UniqueConstraint(columnNames = {"parent_id", "display_order"})
        }
)
public class TestCase {

    @Id
    @GeneratedValue(generator = "uuid2")
    @GenericGenerator(name = "uuid2", strategy = "uuid2")
    @Column(columnDefinition = "VARCHAR(36)", updatable = false)
    private String id;

    @ManyToOne(optional = false)
    @JoinColumn(name = "project_id")
    @JsonBackReference
    private Project project;

    @Column(columnDefinition = "TEXT", nullable = false)
    private String name;

    /**
     * "testcase", "folder", "systemFolder" 중 하나
     */
    @Column(columnDefinition = "TEXT")
    private String type;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(columnDefinition = "TEXT")
    private String preCondition;

    @Column(name = "parent_id")
    private String parentId;

    @ElementCollection
    @CollectionTable(name = "testcasesteps", joinColumns = @JoinColumn(name = "testcase_id"))
    @OrderColumn(name = "step_order")
    private List<TestStep> steps;

    @Column(columnDefinition = "TEXT", name = "expected_results")
    private String expectedResults;

    @Column(name = "display_order", nullable = false)
    private Integer displayOrder = 1;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        if (this.createdAt == null) this.createdAt = LocalDateTime.now();
        this.updatedAt = LocalDateTime.now();
        // displayOrder가 null이면 0이 아니라 1로 초기화
        if (this.displayOrder == null) this.displayOrder = 1;
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }

    @Override
    public String toString() {
        return "TestCase{" +
                "id='" + id + '\'' +
                ", name='" + name + '\'' +
                ", project=" + (project != null ? project.getId() : null) +
                '}';
    }
}
